### ISSUE TITLE
DotProductAttention kernel

### ISSUE THREAD

**Issue by WilliamTambellini** [arrayfire/arrayfire#2787]
This is just a feature request to discuss about the possibility to add a fast optimized DPA (Dot Product Attention) kernel.

https://i2.wp.com/mlexplained.com/wp-content/uploads/2017/12/scaled_dot_product_attention.png?w=840

PyTorch implementation:
https://github.com/pytorch/translate/pull/65/commits/24e58246855fa092e333940956606ef77619cc4e

ToBeDiscussed

**Comment by 9prady9** [comment]
Isn't this a better candidate for https://github.com/arrayfire/arrayfire-ml ?

**Comment by WilliamTambellini** [comment]
Well, I would nt be against reusing afml as long as it is integrated inside the main AF repository (here) to be guarantee to be maintained with uptodate regression tests. 
Short term, adding new ML algos in the ml.hpp is really good enough:
https://github.com/arrayfire/arrayfire/blob/master/include/af/ml.h

**Comment by 9prady9** [comment]
That is true. As long as the implementation is basic/fundamental enough, we can do a function in arrayfire via `ml.h` header. My concern is, from the picture in original description, it seems to be a simple function that can be written on top of existing ArrayFire functions. I could be wrong though, I will investigate a bit on what this `dot product attention` does in detail. I would like to work on it at some point.

**Comment by christophe-murphy** [comment]
This looks like something that could be implemented using existing functions. Perhaps it could be made into an example in examples/machine_learning
